ethereum/spec/src/main/java/tech/pegasys/teku/spec/schemas/registry/BaseSchemaProvider.java

```diff
@@ -0,0 +1,153 @@
+            milestone, isConstant ? lastCreator : lastCreator.withMilestone(milestone));
+      }
+    }
+  }
+
+  @Override
+  public SpecMilestone getEffectiveMilestone(final SpecMilestone milestone) {
+    return getSchemaCreator(milestone).milestone;
+  }
+
+  @Override
+  public T getSchema(final SchemaRegistry registry) {
+    final SpecMilestone milestone = registry.getMilestone();
+    return createSchema(registry, milestone, registry.getSpecConfig());
+  }
+
+  @Override
+  public SchemaId<T> getSchemaId() {
+    return schemaId;
+  }
+
+  protected T createSchema(
+      final SchemaRegistry registry,
+      final SpecMilestone effectiveMilestone,
+      final SpecConfig specConfig) {
+    return getSchemaCreator(effectiveMilestone).creator.apply(registry, specConfig);
+  }
+
+  private SchemaProviderCreator<T> getSchemaCreator(final SpecMilestone milestone) {
+    final SchemaProviderCreator<T> maybeSchemaCreator = milestoneToSchemaCreator.get(milestone);
+    if (maybeSchemaCreator == null) {
+      throw new IllegalArgumentException(
+          "It is not supposed to create a specific version for " + milestone);
+    }
+    return maybeSchemaCreator;
+  }
+
+  @Override
+  public Set<SpecMilestone> getSupportedMilestones() {
+    return milestoneToSchemaCreator.keySet();
+  }
+
+  protected record SchemaProviderCreator<T>(
+      SpecMilestone milestone, BiFunction<SchemaRegistry, SpecConfig, T> creator) {
+
+    private SchemaProviderCreator<T> withMilestone(final SpecMilestone milestone) {
+      return new SchemaProviderCreator<>(milestone, creator);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this).add("milestone", milestone).toString();
+    }
+  }
+
+  static <T> Builder<T> providerBuilder(final SchemaId<T> schemaId) {
+    return new Builder<>(schemaId);
+  }
+
+  static class Builder<T> {
+    private final SchemaId<T> schemaId;
+    final List<SchemaProviderCreator<T>> schemaProviderCreators = new ArrayList<>();
+    private SpecMilestone untilMilestone = SpecMilestone.getHighestMilestone();
+    private boolean isConstant = false;
+
+    private Builder(final SchemaId<T> schemaId) {
+      this.schemaId = schemaId;
+    }
+
+    public Builder<T> withCreator(
+        final SpecMilestone milestone,
+        final BiFunction<SchemaRegistry, SpecConfig, T> creationSchema) {
+      checkArgument(
+          schemaProviderCreators.isEmpty()
+              || milestone.isGreaterThan(schemaProviderCreators.getLast().milestone),
+          "Creator's milestones must added in strict ascending order for %s",
+          schemaId);
+
+      schemaProviderCreators.add(new SchemaProviderCreator<>(milestone, creationSchema));
+      return this;
+    }
+
+    public Builder<T> until(final SpecMilestone untilMilestone) {
+      this.untilMilestone = untilMilestone;
+      return this;
+    }
+
+    public Builder<T> constant(final boolean isConstant) {
```

Comment:
Nit: I'd expose 2 non-param lazy shortcuts instead, `constant()` and `nonConstant()`, or just single `constant()` as it's useless to call it 2 times and default is nonConstant